### PR TITLE
add individual sigmoid masks to Boundless DAS

### PR DESCRIPTION
**Pull request by jiudingsun01**
Add new trainable intervention class SigmoidMaskRotatedSpaceIntervention()

### CODE REVIEW

File: models/interventions.py [hunk 1]
```diff
@@ -253,6 +253,58 @@ def __str__(self):
...
+    def set_interchange_dim(self, interchange_dim):
+        """interchange dim is learned and can not be set"""
+        assert False
+
+    def forward(self, base, source, subspace=None):
+        batch_size = base.shape[0]
+        rotated_base = self.rotate_layer(base)
+        rotated_source = self.rotate_layer(source)
+        # get boundary
+        intervention_boundaries = torch.clamp(
```
Comment (frankaging) [inline models/interventions.py:291]: i think you can remove this boundary as the masked version does not require this.

File: models/interventions.py [hunk 2]
```diff
@@ -253,6 +253,58 @@ def __str__(self):
...
+            self.masks, 1e-3, 1)
+        boundary_mask = torch.sigmoid(self.masks / self.temperature)
+        
+        boundary_mask = torch.ones(
```
Comment (frankaging) [inline models/interventions.py:295]: you can avoid doing this as the mask should be already bound between 0 and 1 after sigmoid.

File: models/interventions.py [hunk 3]
```diff
@@ -253,6 +253,58 @@ def __str__(self):
...
+            batch_size, device=base.device).unsqueeze(dim=-1)*boundary_mask
+        boundary_mask = boundary_mask.to(rotated_base.dtype)
+        # interchange
+        rotated_output = (1. - boundary_mask)*rotated_base + boundary_mask*rotated_source
+        # inverse output
+        output = torch.matmul(rotated_output, self.rotate_layer.weight.T)
+        return output.to(base.dtype)
+
+    def __str__(self):
+        return f"BoundlessRotatedSpaceIntervention(embed_dim={self.embed_dim})"
```
Comment (frankaging) [inline models/interventions.py:305]: nit: change the name here as well

File: models/interventions.py [hunk 4]
```diff
@@ -253,6 +253,58 @@ def __str__(self):
...
+    def __init__(self, embed_dim, **kwargs):
+        super().__init__(**kwargs)
+        rotate_layer = RotateLayer(embed_dim)
+        self.rotate_layer = torch.nn.utils.parametrizations.orthogonal(
+            rotate_layer)
+        self.subspace_partition = kwargs["subspace_partition"] \
+            if "subspace_partition" in kwargs else None
+        # TODO: in case there are subspace partitions, we
+        #       need to initialize followings differently.
+        self.masks = torch.nn.Parameter(
```
Comment (frankaging) [inline models/interventions.py:268]: it maybe better to initialize this such that all bits in the mask are turned on initially. the idea is that we want to push down the number of bits being turned on through an external loss such as `mask.sum()`.

**Review by frankaging — CHANGES_REQUESTED** [review]
LGTM otherwise, please address the comments.
